### PR TITLE
Error message improvements for unions with identical discriminants

### LINKED ISSUE #40934: Union with non-distinct discriminating property results in confusing error messages
**TypeScript Version:**  4.0.2

**Search Terms:** discriminated union, error messages

**Code**

```ts
type Action = {
    type: 'action1';
    payload: {
        property1: string;
    };
} | {
    type: 'action1';
    payload: {
        property2: number;
    };
} | {
    type: 'action2';
    payload: {
        property3: boolean;
    };
}

const action: Action = { // Actual error is here
    type: 'action1',
    payload: {
        property3: true // Expected error here
    }
}
```

**Expected behavior:**

Expected error at `action.payload.property3` declaration, perhaps something like:

```
Object literal may only specify known properties, and 'property3' does not exist in type '{ property1: string; } | { property2: number; }'.
```

**Actual behavior:**

```
Type '{ type: "action1"; payload: { property3: true; }; }' is not assignable to type 'Action'.
  Type '{ type: "action1"; payload: { property3: true; }; }' is not assignable to type '{ type: "action2"; payload: { property3: boolean; }; }'.
    Types of property 'type' are incompatible.
      Type '"action1"' is not assignable to type '"action2"'.(2322)
```

**Playground Link:** https://www.typescriptlang.org/play?#code/C4TwDgpgBAggxsAlgewHZQLxQN4CgoFSiQBcUA5AIYIqoCM5A3PoWJSADbKUAmZehQVDAAnZJBGg6ZAM7ARiVAHNmggL7M1UAD44WBYhDJUaaBqtbsuvfvqGjxESSABMZVAFcAtgCMnFgg1cLV0BQkNjaiQ0FyY7Nk5uPj0hVjEJUABmMh9kZA4ISlQAqCC1XFw4NDkoKNoyeGj0LDCDcCMKOrNyABp4qyTbVIIHDJBsohEPCDtytSA

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This PR would fix #40934, but we will be changing the core logic used in #42556.

However, I'm opening this because I wanted to point out the sorts of ~improvements~ changes we can see though if we're willing do a little extra work to track multiple object types with identical discriminants though. That logic can possibly be incorporated into #42556.

### CODE REVIEW

File: tests/baselines/reference/contextualTypeShouldBeLiteral.errors.txt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        type2: 'y',
+        ~~~~~~~~~~
+!!! error TS2345: Argument of type '{ type2: "y"; value: "done"; method(): void; }' is not assignable to parameter of type 'X2 | Y2'.
+!!! error TS2345:   Object literal may only specify known properties, but 'type2' does not exist in type 'X2'. Did you mean to write 'type1'?
+        value: 'done',
+        method() {
+            this;
+            this.value;
+        }
+    });
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/contextualTypeShouldBeLiteral.errors.txt:54]: This is a new error - I'd argue this is actually a regression.

File: tests/baselines/reference/excessPropertyCheckWithMultipleDiscriminants.errors.txt [hunk 2]
```diff
@@ -63,6 +65,9 @@ tests/cases/compiler/excessPropertyCheckWithMultipleDiscriminants.ts(57,5): erro
         p2: true,
         p3: 42,
         p4: "hello"
+        ~~~~~~~~~~~
+!!! error TS2322: Type '{ p1: "left"; p2: true; p3: number; p4: string; }' is not assignable to type 'DisjointDiscriminants'.
+!!! error TS2322:   Object literal may only specify known properties, and 'p4' does not exist in type '{ p1: "left"; p2: true; p3: number; }'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/excessPropertyCheckWithMultipleDiscriminants.errors.txt:70]: This one is definitely an improvement.

File: tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt [hunk 3]
```diff
@@ -71,33 +71,36 @@ tests/cases/compiler/excessPropertyCheckWithUnions.ts(114,63): error TS2322: Typ
     // no error for ambiguous tag, even when it could satisfy both constituents at once
     amb = { tag: "A", x: "hi" }
     amb = { tag: "A", y: 12 }
+                      ~~~~~
+!!! error TS2322: Type '{ tag: "A"; y: number; }' is not assignable to type 'Ambiguous'.
+!!! error TS2322:   Object literal may only specify known properties, and 'y' does not exist in type '{ tag: "A"; x: string; }'.
     amb = { tag: "A", x: "hi", y: 12 }
+                               ~~~~~
+!!! error TS2322: Type '{ tag: "A"; x: string; y: number; }' is not assignable to type 'Ambiguous'.
+!!! error TS2322:   Object literal may only specify known properties, and 'y' does not exist in type '{ tag: "A"; x: string; }'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt:80]: These are...ambiguous! Unclear if they should error.

File: tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt [hunk 4]
```diff
@@ -71,33 +71,36 @@ tests/cases/compiler/excessPropertyCheckWithUnions.ts(114,63): error TS2322: Typ
...
     
     // correctly error on excess property 'extra', even when ambiguous
     amb = { tag: "A", x: "hi", extra: 12 }
                                ~~~~~~~~~
 !!! error TS2322: Type '{ tag: "A"; x: string; extra: number; }' is not assignable to type 'Ambiguous'.
-!!! error TS2322:   Object literal may only specify known properties, and 'extra' does not exist in type 'Ambiguous'.
+!!! error TS2322:   Object literal may only specify known properties, and 'extra' does not exist in type '{ tag: "A"; x: string; }'.
     amb = { tag: "A", y: 12, extra: 12 }
-                             ~~~~~~~~~
+                      ~~~~~
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt:88]: Regression

File: tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt [hunk 5]
```diff
@@ -71,33 +71,36 @@ tests/cases/compiler/excessPropertyCheckWithUnions.ts(114,63): error TS2322: Typ
...
+!!! error TS2322:   Property 'x' is missing in type '{ tag: "A"; }' but required in type '{ tag: "A"; x: string; }'.
+!!! related TS2728 tests/cases/compiler/excessPropertyCheckWithUnions.ts:16:5: 'x' is declared here.
     amb = { tag: "A", z: true }
-    ~~~
+                      ~~~~~~~
 !!! error TS2322: Type '{ tag: "A"; z: true; }' is not assignable to type 'Ambiguous'.
-!!! error TS2322:   Type '{ tag: "A"; z: true; }' is not assignable to type '{ tag: "B"; z: boolean; }'.
-!!! error TS2322:     Types of property 'tag' are incompatible.
-!!! error TS2322:       Type '"A"' is not assignable to type '"B"'.
+!!! error TS2322:   Object literal may only specify known properties, and 'z' does not exist in type '{ tag: "A"; x: string; }'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt:103]: This one is sort of an improvement - ideally, we'd be able to say that `z` was in another branch.

**Review by DanielRosenwasser — COMMENTED** [review]
Looking at this again, it's way more arguable. I think part of the approach that might be failing is that we're stopping at the first type, and not continuously refining the union of types given to us.

File: tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt (hunk 3 above)
Comment (andrewbranch) [reply tests/baselines/reference/excessPropertyCheckWithUnions.errors.txt:80]: Wait, why should they be errors? Particularly `amb = { tag: "A", y: 12 }` seems super wrong. But this one too seems like excess property checking is a little overzealous. For non-discriminated unions of objects, excess property checking only complains about properties non present in any constituent. Here, I would think the same rule would apply after we have filtered out the B and C constituents.

File: tests/baselines/reference/contextualTypeShouldBeLiteral.errors.txt (hunk 1 above)
Comment (andrewbranch) [reply tests/baselines/reference/contextualTypeShouldBeLiteral.errors.txt:54]: Yeah, isn’t this perfectly assignable to `Y2`?

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by sandersn** [comment]
This experiment is pretty old, so I'm going to close it to reduce the number of open PRs.

**Pull request closed without merging**